ports/zephyr/zephyr_storage.c

```diff
@@ -0,0 +1,268 @@
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 NXP
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "zephyr_storage.h"
+#include "py/runtime.h"
+
+#if MICROPY_VFS
+#include "extmod/vfs.h"
+#endif
+
+#ifdef CONFIG_DISK_ACCESS
+#include <disk/disk_access.h>
+#endif
+
+#ifdef CONFIG_FLASH_MAP
+#include <storage/flash_map.h>
+#endif
+
+#ifdef CONFIG_DISK_ACCESS
+typedef struct _zephyr_disk_access_obj_t {
+    mp_obj_base_t base;
+    const char *pdrv;
+    int block_size;
+    int block_count;
+} zephyr_disk_access_obj_t;
+
+STATIC void zephyr_disk_access_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_disk_access_obj_t *self = self_in;
+    mp_printf(print, "DiskAccess(%s)", self->pdrv);
+}
+
+STATIC mp_obj_t zephyr_disk_access_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    zephyr_disk_access_obj_t *self = m_new_obj(zephyr_disk_access_obj_t);
+    self->base.type = type;
+    self->pdrv = mp_obj_str_get_str(args[0]);
+
+    if (disk_access_init(self->pdrv) != 0) {
+        mp_raise_ValueError("disk not found");
+    }
+
+    if (disk_access_ioctl(self->pdrv, DISK_IOCTL_GET_SECTOR_SIZE, &self->block_size)) {
+        mp_raise_ValueError("unable to get sector size");
+    }
+
+    if (disk_access_ioctl(self->pdrv, DISK_IOCTL_GET_SECTOR_COUNT, &self->block_count)) {
+        mp_raise_ValueError("unable to get block count");
+    }
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_obj_t zephyr_disk_access_readblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    zephyr_disk_access_obj_t *self = self_in;
+    mp_buffer_info_t bufinfo;
+    int ret;
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+    ret = disk_access_read(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return MP_OBJ_NEW_SMALL_INT(ret);
+}
+
```

Comment:
we usually don't put an empty line between the function and the wrapper just below, to indicate that they have a very close relationship